toolchain/sem_ir/function.cpp

```diff
@@ -28,19 +29,22 @@ auto GetCalleeFunction(const File& sem_ir, InstId callee_id) -> CalleeFunction {
     callee_id = bound_method->function_decl_id;
   }
 
-  // Identify the function we're calling.
+  // Identify the function we're calling by its type.
   auto val_id = sem_ir.constant_values().GetConstantInstId(callee_id);
   if (!val_id.has_value()) {
     return result;
   }
-  auto val_inst = sem_ir.insts().Get(val_id);
-  auto struct_val = val_inst.TryAs<StructValue>();
-  if (!struct_val) {
-    result.is_error = val_inst.type_id() == SemIR::ErrorInst::SingletonTypeId;
-    return result;
+  auto fn_type_inst =
+      sem_ir.types().GetAsInst(sem_ir.insts().Get(val_id).type_id());
+
+  if (auto impl_fn_type = fn_type_inst.TryAs<ImplFunctionType>()) {
+    result.self_type_id = impl_fn_type->self_id;
```

Comment:
```suggestion
    // Combine the associated function's `self` with the interface function data.
    result.self_type_id = impl_fn_type->self_id;
```
Suggesting a comment I think might've helped me see the control flow more easily.